src/Disks/DiskLocal.cpp

```diff
@@ -59,28 +61,29 @@ class DiskLocalReservation : public IReservation
 class DiskLocalDirectoryIterator : public IDiskDirectoryIterator
 {
 public:
-    explicit DiskLocalDirectoryIterator(const String & disk_path_, const String & dir_path_)
-        : dir_path(dir_path_), iter(disk_path_ + dir_path_)
+    explicit DiskLocalDirectoryIterator(const fs::path & disk_path_, const String & dir_path_)
+        : dir_path(dir_path_), entry(disk_path_ / dir_path_)
     {
     }
 
-    void next() override { ++iter; }
+    void next() override { ++entry; }
 
-    bool isValid() const override { return iter != Poco::DirectoryIterator(); }
+    bool isValid() const override { return entry != fs::directory_iterator(); }
 
     String path() const override
     {
-        if (iter->isDirectory())
-            return dir_path + iter.name() + '/';
+        if (entry->is_directory())
+            return dir_path / entry->path().filename() / "";
```

Comment:
No, not equivalent because same path can be passed to multiple disks, see https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeData.cpp#L4324